src/multisig/multisig_kex.cpp

```diff
@@ -0,0 +1,638 @@
+   *        - If msg is not for the last round, then these derivations are also stored in the output message
+   *          so they can be sent to other participants, who will make more DH derivations for the next kex round.
+   *        - If msg is for the last round, then these derivations won't be sent to other participants.
+   *          Instead, they are converted to share secrets (i.e. s = H(derivation)) and multiplied by G.
+   *          The keys s*G are sent to other participants in the message, so they can be used to produce the final
+   *          multisig key via generate_multisig_spend_public_key().
+   *            - The values s are the local account's shares of the final multisig key's private key. The caller can
+   *              compute those values with calculate_multisig_keypair_from_derivation() (or compute them directly).
+   * return: multisig kex message for the specified round
+   */
+  //----------------------------------------------------------------------------------------------------------------------
+  static multisig_kex_msg multisig_kex_make_next_msg(const crypto::secret_key &base_privkey,
+    const std::uint32_t round,
+    const std::uint32_t threshold,
+    const std::uint32_t num_signers,
+    const std::unordered_map<crypto::public_key_memsafe, std::unordered_set<crypto::public_key>> &pubkey_origins_map,
+    std::unordered_map<crypto::public_key_memsafe, std::unordered_set<crypto::public_key>> &derivation_origins_map_out)
+  {
+    CHECK_AND_ASSERT_THROW_MES(num_signers > 1, "Must be at least one other multisig signer.");
+    CHECK_AND_ASSERT_THROW_MES(num_signers <= config::MULTISIG_MAX_SIGNERS,
+      "Too many multisig signers specified (limit = 16 to prevent dangerous combinatorial explosion during key exchange).");
+    CHECK_AND_ASSERT_THROW_MES(num_signers >= threshold,
+      "Multisig threshold may not be larger than number of signers.");
+    CHECK_AND_ASSERT_THROW_MES(threshold > 0, "Multisig threshold must be > 0.");
+    CHECK_AND_ASSERT_THROW_MES(round > 1, "Round for next msg must be > 1.");
+    CHECK_AND_ASSERT_THROW_MES(round <= multisig_kex_rounds_required(num_signers, threshold),
+      "Trying to make key exchange message for an invalid round.");
+
+    // make shared secrets with input pubkeys
+    std::vector<crypto::public_key> msg_pubkeys;
+    msg_pubkeys.reserve(pubkey_origins_map.size());
+    derivation_origins_map_out.clear();
+
+    for (const auto &pubkey_and_origins : pubkey_origins_map)
+    {
+      // D = 8 * k_base * K_pubkey
+      // note: must be mul8 (cofactor), otherwise it is possible to leak to a malicious participant if the local
+      //       base_privkey is a multiple of 8 or not
+      // note2: avoid making temporaries that won't be memwiped
+      rct::key derivation_rct;
+      rct::scalarmultKey(derivation_rct, rct::pk2rct(pubkey_and_origins.first), rct::sk2rct(base_privkey));
+      rct::scalarmultKey(derivation_rct, derivation_rct, rct::EIGHT);
```

Comment:
Why?

Reply:
An attacker could still test whether the secret key is a multiple of 2, 4 or 8 (assuming no other checks).

If `pubkey_and_origins.first` is in a subgroup, a multiplication by the subgroup order puts `derivation_rct` onto the primary group, and all other values keep it in the subgroup. So the mul8 afterward would either be putting `derivation_rct` into the primary group for the first time or be performing a mul8 in the primary group.

If the order is flipped, `derivation_rct` is in the primary group (even without other checks) before multiplying by the secret-key.

Reply:
Curve operations are homomorphic...

Suppose `pubkey K = A + X`, where `X` is a point in the cofactor's subgroup. Let `p = privkey`.

```
p K = p A + p X
8 (p K) = 8*p A + 8*p X
            = 8*p A + p*(8 X)
            = 8*p A + p*(I)
            = 8*p A

8 K = 8 A + 8 X
       = 8 A
p (8 K) = 8*p A
```

Same thing.

Reply:
Your rebuttal is irrelevant to the issue because it doesn't address my criticism. (Assuming no other checks), this is leaking information about the secret scalar.

Or perhaps I should ask this - why is the mul8 performed at all here? Typically it is done to "clear" any torsion points on ed25519, and the comments seem to indicate that as well. This clearing should be done before the secret multiplication.

Reply:
I am very confused. Can you show me, with pseudo-algebra, what you are talking about?

Reply:
Oops, this can be closed. Went back through relevant papers - these points I had in mind have orders of `2*l`, `4*l` or `8*l`. Its non-obvious how the torsion points interact with other points, but it makes more sense from a group theory perspective as the commutative property of multiplication still holds.

There's no check for the identity/zero point here, but is tested indirectly when trying to create new kex message.